homeassistant/components/fan/xiaomi_miio.py

```diff
@@ -294,70 +602,235 @@ def speed(self):
+        """Set the extra features."""
+        if self._device_features & FEATURE_SET_EXTRA_FEATURES == 0:
+            return
+
+        await self._try_command(
+            "Setting the extra features of the miio device failed.",
+            self._device.set_extra_features, features)
+
+    async def async_reset_filter(self):
+        """Reset the filter lifetime and usage."""
+        if self._device_features & FEATURE_RESET_FILTER == 0:
+            return
+
+        await self._try_command(
+            "Resetting the filter lifetime of the miio device failed.",
+            self._device.reset_filter)
+
+
+class XiaomiAirHumidifier(XiaomiGenericDevice):
+    """Representation of a Xiaomi Air Humidifier."""
+
+    def __init__(self, name, device, model, unique_id):
+        """Initialize the plug switch."""
+        super().__init__(name, device, model, unique_id)
+
+        if self._model == MODEL_AIRHUMIDIFIER_CA:
+            self._device_features = FEATURE_FLAGS_AIRHUMIDIFIER_CA
+            self._state_attrs.update({attribute: None for attribute in
+                                      AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_CA})
+        else:
+            self._device_features = FEATURE_FLAGS_AIRHUMIDIFIER
+            self._state_attrs.update({attribute: None for attribute in
+                                      AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER})
+
+    async def async_update(self):
+        """Fetch state from the device."""
+        from miio import DeviceException
+
+        # On state change the device doesn't provide the new state immediately.
+        if self._skip_update:
+            self._skip_update = False
+            return
+
+        try:
+            state = await self.hass.async_add_job(
+                self._device.status)
+            _LOGGER.debug("Got new state: %s", state)
+
+            self._available = True
+            self._state = state.is_on
+
+            if self._model == MODEL_AIRHUMIDIFIER_CA:
+                self._state_attrs.update(
+                    {key: self._extract_value_from_attribute(state, value)
+                     for key, value in
+                     AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_CA.items()})
+            else:
+                self._state_attrs.update(
+                    {key: self._extract_value_from_attribute(state, value)
+                     for key, value in
+                     AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER.items()})
+
+        except DeviceException as ex:
+            self._available = False
+            _LOGGER.error("Got exception while fetching the state: %s", ex)
+
+    def speed_list(self) -> list:
+        """Get the list of available speeds."""
+        from miio.airpurifier import OperationMode
+        if self._model == MODEL_AIRHUMIDIFIER_V1:
+            return [mode.name for mode in OperationMode if mode.name != 'Auto']
+
+        # Air Humidifier CA
+        return [mode.name for mode in OperationMode]
```

Comment:
Is this also valid for non-CA humidifiers? If yes, the comment should be updated.

Reply:
I'm using the V1 as new default now. The CA is a very specific device version. My approach was to provide all known operation modes to reduce it later on for specific devices.